File: modules/swagger-jaxrs/src/test/scala/ReferenceTest.scala
import com.wordnik.swagger.converter.ModelConverters
import models.Pet
import resources._
import matchers.SerializationMatchers._

import org.scalatest.{FlatSpec, Matchers}
import com.wordnik.swagger.models.properties.{RefProperty, MapProperty}

import com.wordnik.swagger.models.Swagger
import com.wordnik.swagger.jaxrs.Reader

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FlatSpec
import org.scalatest.Matchers

@RunWith(classOf[JUnitRunner])
class ReferenceTest extends FlatSpec with Matchers {

  it should "Scan a model with common reference and reference with ApiModel" in {
    val props = ModelConverters.getInstance().readAll(classOf[Pet]).get("Pet").getProperties()
    val category = props.get("category").asInstanceOf[RefProperty]
    category.getType() should be ("ref")
    category.get$ref() should be ("#/definitions/Category")

    val categoryWithApiModel = props.get("categoryWithApiModel").asInstanceOf[RefProperty]
    categoryWithApiModel.getType() should be ("ref")
    categoryWithApiModel.get$ref() should be ("#/definitions/MyCategory")
  }

  it should "Scan API with operation and response references" in {
    val reader = new Reader(new Swagger())
    val swagger = reader.read(classOf[ResourceWithReferences])

    swagger should serializeToJson (
      """{
  "swagger": "2.0",
  "tags": [
    {
      "name": "basic"
    }
  ],
  "paths": {
    "/anotherModel": {
      "get": {
        "tags": [
          "basic"
        ],
        "summary": "Get Another Model",
        "description": "",
        "operationId": "getAnotherModel",
        "parameters": [],
        "responses": {
          "200": {
            "description": "successful operation",
            "schema": {
              "$ref": "http://swagger.io/schemas.json#/Models"
            }
          }
        }
      }
    },
    "/model": {
      "get": {
        "tags": [
          "basic"
        ],
        "summary": "Get Model",
        "description": "",
        "operationId": "getModel",
        "parameters": [],
        "responses": {
          "200": {
            "description": "successful operation",
            "schema": {
              "$ref": "#/definitions/ModelContainingModelWithReference"
            }
          }
        }
      }
    },
    "/some": {
      "get": {
        "tags": [
          "basic"
        ],
        "summary": "Get Some",
        "description": "",
        "operationId": "getSome",
        "parameters": [],
        "responses": {
          "200": {
            "description": "successful operation",
            "schema": {
              "$ref": "http://swagger.io/schemas.json#/Models/SomeResponse"
            }
          }
        }
      }
    },
    "/test": {
      "get": {
        "tags": [
          "basic"
        ],
        "operationId": "getTest",
        "parameters": [],
        "responses": {
          "500": {
            "description": "Error",
            "schema": {
              "$ref": "http://swagger.io/schemas.json#/Models/ErrorResponse"
            }
          }
        }
      }
    },
    "/testSome": {
      "get": {
        "tags": [
          "basic"
        ],
        "summary": "Get Some",
        "description": "",
        "operationId": "getTestSome",
        "parameters": [],
        "responses": {
          "200": {
            "description": "successful operation",
            "schema": {
              "$ref": "http://swagger.io/schemas.json#/Models/SomeResponse"
            }
          },
          "500": {
            "description": "Error",
            "schema": {
              "$ref": "http://swagger.io/schemas.json#/Models/ErrorResponse"
            }
          }
        }
      }
    }
  },
  "definitions": {
    "ModelWithReference": {
      "type": "object",
      "properties": {
        "description": {
          "$ref": "http://swagger.io/schemas.json#/Models/Description"
        }
      }
    },
    "ModelContainingModelWithReference": {
      "type": "object",
      "properties": {
        "model": {
          "$ref": "http://swagger.io/schemas.json#/Models"
        },
        "anotherModel": {
          "$ref": "http://swagger.io/schemas.json#/Models/AnotherModel"
        }
      }
    }
  }
}""")
  }
}

File: modules/swagger-jaxrs/src/test/java/io/swagger/ReferenceTest.java
package io.swagger;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.swagger.converter.ModelConverters;
import io.swagger.jaxrs.Reader;
import io.swagger.matchers.SerializationMatchers;
import io.swagger.models.Pet;
import io.swagger.models.Swagger;
import io.swagger.models.properties.Property;
import io.swagger.models.properties.RefProperty;
import io.swagger.models.refs.GenericRef;
import io.swagger.resources.ResourceWithMoreReferences;
import io.swagger.resources.ResourceWithReferences;
import io.swagger.util.Json;
import io.swagger.util.ReferenceSerializationConfigurer;
import io.swagger.util.ResourceUtils;
import org.testng.annotations.Test;

import java.io.IOException;
import java.util.Map;

import static org.testng.Assert.assertEquals;

public class ReferenceTest {

    @Test(description = "Scan a model with common reference and reference with ApiModel")
    public void scanModel() {
        final Map<String, Property> props = ModelConverters.getInstance().readAll(Pet.class).get("Pet").getProperties();
        final RefProperty category = (RefProperty) props.get("category");
        assertEquals(category.getType(), "ref");
        assertEquals(category.get$ref(), "#/definitions/Category");

        final RefProperty categoryWithApiModel = (RefProperty) props.get("categoryWithApiModel");
        assertEquals(categoryWithApiModel.getType(), "ref");
        assertEquals(categoryWithApiModel.get$ref(), "#/definitions/MyCategory");
    }

    @Test(description = "Scan API with operation and response references")
    public void scanAPI() throws IOException {

        final Swagger swagger = new Reader(new Swagger()).read(ResourceWithReferences.class);
        final String json = ResourceUtils.loadClassResource(getClass(), "ResourceWithReferences.json");
        SerializationMatchers.assertEqualsToJson(swagger, json);
    }

    @Test(description = "Scan API with references")
    public void scanRef() throws IOException {

        final Swagger swagger = new Reader(new Swagger()).read(ResourceWithMoreReferences.class);
        final String json = ResourceUtils.loadClassResource(getClass(), "ResourceWithMoreReferences.json");
        SerializationMatchers.assertEqualsToJson(swagger, json);
    }

    @Test(description = "Serialize API with references and OriginalRefMixin activated")
    public void serializeRefWithOriginalRef() throws Exception {

        // workaround for https://github.com/FasterXML/jackson-databind/issues/1998
        ObjectMapper mapper = Json.mapper().copy();
        ReferenceSerializationConfigurer.serializeAsOriginalRef(mapper);

        final String json = ResourceUtils.loadClassResource(getClass(), "ResourceWithMoreReferencesAsOriginalRef.json");
        Swagger swagger = Json.mapper().readValue(json, Swagger.class);
        SerializationMatchers.assertEqualsToString(swagger, json, mapper);
    }

    @Test(description = "Serialize API with references and internal ref also with dots activated")
    public void serializeRefWithInternalRef() throws Exception {
        try {
            GenericRef.internalRefWithAnyDot();
            final String json = ResourceUtils.loadClassResource(getClass(), "ResourceWithMoreReferencesWithInternalRef.json");
            Swagger swagger = Json.mapper().readValue(json, Swagger.class);
            SerializationMatchers.assertEqualsToJson(swagger, json);
        } finally {
            GenericRef.relativeRefWithAnyDot();
        }
    }

    @Test(description = "Scan API with references and OriginalRefMixin activated")
    public void scanRefWithOriginalRef() throws IOException {

        // workaround for https://github.com/FasterXML/jackson-databind/issues/1998
        ObjectMapper mapper = Json.mapper().copy();
        ReferenceSerializationConfigurer.serializeAsOriginalRef(mapper);

        final Swagger swagger = new Reader(new Swagger()).read(ResourceWithMoreReferences.class);
        final String json = ResourceUtils.loadClassResource(getClass(), "ResourceWithMoreReferencesAsOriginalRef.json");
        SerializationMatchers.assertEqualsToString(swagger, json, mapper);
    }

}
